File: cc-algo/c/array/searchInsert.c
// 复杂度 O(n)
int searchInsert(int* nums, int numsSize, int target) {
    for (int i = 0; i < numsSize; i++) {
        if (target <= nums[i]) {
            return i;
        }
    }
    return numsSize;
}

int searchInsertBinary(int* nums, int numsSize, int target) {
    if (nums[numsSize - 1] <= target) {
        return numsSize;
    }
    int left = 0, right = numsSize - 1;
    while (left < right) {
        int mid = left + (left + right) / 2;
        if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

File: go-algo/algo/arrays/search_insert.go
package arrays

// 二分法搜索
// 定位
func searchInsert(nums []int, target int) int {
	n := len(nums)
	left, right := 0, n-1
	ans := n
	for left <= right {
		mid := (right-left)>>1 + left
		if target <= nums[mid] {
			ans = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return ans
}

// 二分法搜索
func searchBinary(nums []int, target int) int {
	ans := -1
	left, right := 0, len(nums)-1 // 1 首尾
	for left <= right {           // 2
		mid := (right-left)>>1 + left
		if target < nums[mid] {
			right = mid - 1 // 3
		} else if target > nums[mid] {
			left = mid + 1 // 4
		} else {
			ans = mid
			break
		}
	}
	return ans
}

func searchRange(nums []int, target int) []int {
	posRange := []int{-1, -1}
	posLeft := searchInsert(nums, target)
	posRight := searchInsert(nums, target+1)
	if posLeft >= len(nums) || nums[posLeft] != target {
		return posRange
	}
	return []int{posLeft, posRight - 1}
}
